flower/utils/results/stores/database.py

```diff
@@ -1,13 +1,107 @@
+                name=task.name,
+            )
+
+        return Result(
+            task_id=task.task_id,
+            status=task.status,
+            date_done=task.date_done,
+            result=task.result,
+            traceback=task.traceback,
+        )
+
+    def deserialize_binary_column_value(self, value: bytes) -> Any:
+        """
+        Attempt to deserialize the provided `value` using the available serialization decoders. Celery stores task
+        `args` and `kwargs` in binary columns, but the proper decoding mechanism for those binary columns is not
+        immediately obvious. These fields get serialized bsed on whatever the value of `Celery.conf.result_serializer`
+        is at the time the task result is saved. However, it's possible that the value of that config setting changed
+        across different Celery processes, and therefore we may be dealing with a database that has co-mingled records
+        from different serializers. Unfortunately, there is no column in the database schema that records which
+        serializer was used for each task.
+
+        To work around this limitation, this method takes guesses at the serialization of `value` based on whatever
+        serializers are available in the active `result_accept_content` or `accept_content` Celery config setting.
+        Each serializer will attempt deserialization, and if one succeeds, we return the deserialized value immediately.
+        If all deserialization attempts fail, we will gracefully return the original bytes value.
+
+        TODO: currently this method only attempts deserialization with JSON and pickle. We should support more built-in
+          content types, and potentially allow for deserialization using custom encodings. We chose to limit the
+          supported serializers here because JSON and pickle are the only serialization mechanisms available without
+          additional dependencies (e.g. 'yaml' requires the inclusion of the third-party `yaml` library).
+        """
+        celery_result_accept_content: list[str] = (
+            self.backend.app.conf.result_accept_content
+            or self.backend.app.conf.accept_content
+        )
+        accept_content_types: list[str] = [item.lower() for item in celery_result_accept_content]
+
+        if 'json' in accept_content_types or 'application/json' in accept_content_types:
+            try:
+                return kombu.serialization.registry._decoders['application/json'](value)
+            except (json.JSONDecodeError, UnicodeDecodeError):
+                pass
+
+        if 'pickle' in accept_content_types or 'application/x-python-serialize' in accept_content_types:
+            try:
+                return kombu.serialization.registry._decoders['application/x-python-serialize'](value)
+            except pickle.UnpicklingError:
+                pass
+
+        # couldn't deserialize; just fall back to the original byte string
+        return value
```

Comment:
This feels less useful to me than a default value like "could not deserialize content type whatever/whatever", or some sort of in-your-face informational method that clearly described the problem

Reply:
Hm. I kind of like it because, with human-readable serialization methods like JSON, you get something pretty readable. Instead of the properly-deserialized `{"key": "value"}`, you would see `b'{"key": "value"}`. As a user I'd rather run into that than completely lack visibility into what the value is.

Reply:
It’s a difference of whether or not you want to paper over the problem, my advice is don’t paper over it to start in order to psychologically influence fixes. If you unintentionally hamper it, it becomes a “bug” vs “enhancement”, a bug being more likely to getting resolved and easier to diagnose.

Reply:
> it becomes a “bug” vs “enhancement”, a bug being more likely to getting resolved and easier to diagnose.

Yeah this is a really good point. I'll make the change to still return gracefully, but with some `<could not deserialize ...>` error string. I want to make sure that our list views can still render even if there's one bad apple, so raising an exception would be too heavy-handed. But I suppose an error string would force us to fix the bug.

Reply:
Typing out that last comment made me think of something. Because of the way we have Flower set up as a standalone image, it won't be able to unpickle custom classes. If you have a class `app.utils.MyThing()` in tempus, and Flower encounters this in `args`, it will fail unpickling because `app.utils.MyThing()` isn't known to Flower. 

Perhaps this is another argument in favor of more tightly coupling Flower with the application that's using it. The way Flower gets invoked is `celery -A ... flower`, with the `flower` portion being a subcommand. We could pretty easily include `shipwell/flower` as a dependency to tempus, then have a tempus ECS task very similar to our celery worker, but that instead runs `celery -A ... flower`. It would have access to the tempus codebase, so it could unpickle without issue. 

CC @dylan-shipwell

Reply:
great call out

> [...] it will fail unpickling because `app.utils.MyThing()` isn't known to Flower.
> 
so, is pickle-as-a-serializer an implementation detail of celery/flower that we don't control?

cause if it's an implementation detail, that would suggest to me that we should only be passing stdlib types like string to celery; serializing our sw data first to a string-safe format like json/xml/toml/yaml/sql/whatever

either way
- we should adapt sw to use celery safe types, use the neutral interface of stdlib types only (dict/tuples/lists of str and int, or one giant string like a json payload)
- or we should adapt celery to be transparent about sw types, by using our own celery middleware interface that will for example pickle-in-pickle sw classes so celery never needs to know about our implementation details end-to-end

the former is probably easier to get visibility on because humans can mostly read json but mostly can't read pickle or pickle in pickle with ease, but the latter might be nicer to make celery transparent to shipwell's existing code.
🤷 

> [... maybe] include `shipwell/flower` as a dependency to tempus, then have a tempus ECS task very similar to our celery worker, but that instead runs `celery -A ... flower`. It would have access to the tempus codebase, so it could unpickle without issue.

I would avoid <s>cohereing</s> oops i mean [coupling](https://en.wikipedia.org/wiki/Cohesion_(computer_science)#:~:text=In%20computer%20programming%2C%20cohesion%20refers,concept%20served%20by%20that%20class.) shipwells/tempus's interfaces and celery's interfaces together. i'd rather these be independ things that are not in scope of each other impacting library upgrades and code developments/rollouts.

it's a little bit funny because it's a round-trip scenario, celery is only the data carrier providing carriage of a customer payload. That combined with it-happens-to-be that both celery and shipwell use python and so our interface could be intermixed and coupled together tightly, but putting flower in scope of needing to understand shipwell types sounds unrealistic to me, we'd never ask AWS or rollbar or sw tennants to be able to deserialize a Shipment class pickle, at best we get strings and json-in-strings as a closet fit. if we ever moved to managed celery, this would be a blocker for us. Go for high cohesion low coupling for all network-transparent services like celery/flower/rest/logs/sql etc.


---
going on too long ...

drawing experience from other binary format i've worked with, 
even something "still-binary" but std-lib safe to decode, 
like base64-encoded-gzip-of-cpickle-with-discrimination-headers would work well,

if i had to make one up, like, this kind of thing would probably be forward-safe to operate
```
pickle.dumps(tuple(
    # "SW_BIN_1" is the eight byte file type discriminator.
    # (dylang) well technically, its double-encoded so it's actually these 4 bytes on py 3.11 b'\x80\x04\x95\x16'
    # but these will ver version-dependent bytes like any magic number 
    # https://github.com/python/cpython/blob/b14f0ab51cb4851b25935279617e388456dcf716/PC/launcher.c#L1271
    # b'\x80\x04\x95\x16\x00\x00\x00\x00\x00\x00\x00\x8c\x05magic\x94C\x08helowrld\x94\x86\x94.' ...
    
    "magic": struct.pack('!ll', int.from_bytes(b'SW_B'), int.from_bytes(b'IN_1')),
    "sw_build_id", SW_META.BUILD_ARTIFACT_ID,
    "sw_release_version", SW_META.RELEASE_ID,
    "cannonical_class_name", f"{__module__.__name__}.{klass.__name__}",
    "base64_gzip_pickle", base64.b64encode(gzip.compress(cpickle.pickle(klass)))
))
```

python gives us a ton of flexibility, we could use a dictionary here but ordering is not historically guaranteed with dicts, so a tuple makes the header bytes and headers predictable such that any naive user could identify what data is in this packet.

Reply:
> so, is pickle-as-a-serializer an implementation detail of celery/flower that we don't control?

If we choose to use pickle in Celery, then it operates the same way that any usage of Python's `pickle` library operates. Which is kinda like `java.io.Serializable`. So we don't have control over the implementation, but we do have control over the usage. 

> cause if it's an implementation detail, that would suggest to me that we should only be passing stdlib types like string to celery; serializing our sw data first to a string-safe format like json/xml/toml/yaml/sql/whatever

Our friend @phillipuniverse probably agrees with you here; and it's not a bad argument. The usage of `pickle` is always controversial, and in this case there are certainly downsides when it comes to tight coupling. That said, this is a library so we should likely support `pickle` along with other methods. If some Flower user _really_ loves `pickle`, they should be able to use that even if they might end up seeing some ugly byte strings on their webpage.